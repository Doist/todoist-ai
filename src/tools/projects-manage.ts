import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { formatNextSteps } from '../utils/response-builders.js'
import { ToolNames } from '../utils/tool-names.js'

const { SECTIONS_MANAGE, TASKS_ADD_MULTIPLE, PROJECTS_LIST, TASKS_LIST_FOR_CONTAINER, OVERVIEW } =
    ToolNames

const ArgsSchema = {
    id: z
        .string()
        .min(1)
        .optional()
        .describe(
            'The ID of the project to update. If provided, updates the project. If omitted, creates a new project.',
        ),
    name: z.string().min(1).describe('The name of the project.'),
}

const projectsManage = {
    name: ToolNames.PROJECTS_MANAGE,
    description:
        'Add a new project or update an existing project. If id is provided, updates the project. If id is omitted, creates a new project.',
    parameters: ArgsSchema,
    async execute(args, client) {
        if (args.id) {
            // Update existing project
            const project = await client.updateProject(args.id, { name: args.name })
            const textContent = generateTextContent({ project, operation: 'updated' })

            return getToolOutput({
                textContent,
                structuredContent: {
                    project,
                    operation: 'updated',
                },
            })
        }

        // Create new project
        const project = await client.addProject({ name: args.name })
        const textContent = generateTextContent({ project, operation: 'created' })

        return getToolOutput({
            textContent,
            structuredContent: {
                project,
                operation: 'created',
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateTextContent({
    project,
    operation,
}: {
    project: { id: string; name: string }
    operation: 'created' | 'updated'
}): string {
    const action = operation === 'created' ? 'Created' : 'Updated'
    const summary = `${action} project: ${project.name} â€¢ id=${project.id}`

    // Context-aware next steps based on operation
    const nextSteps: string[] = []

    if (operation === 'created') {
        // For new projects, suggest logical setup workflow
        nextSteps.push(`Use ${SECTIONS_MANAGE} to organize this project with sections`)
        nextSteps.push(`Use ${TASKS_ADD_MULTIPLE} to add your first tasks`)
        nextSteps.push(`Use ${OVERVIEW} with projectId to see updated project structure.`)
    } else {
        // For updated projects, suggest review and management
        nextSteps.push(`Use ${OVERVIEW} with projectId=${project.id} to see project structure`)
        nextSteps.push(`Use ${PROJECTS_LIST} to see all projects with updated name`)
        nextSteps.push(`Use ${TASKS_LIST_FOR_CONTAINER} to review existing tasks`)
    }

    const next = formatNextSteps(nextSteps)
    return `${summary}\n${next}`
}

export { projectsManage }
