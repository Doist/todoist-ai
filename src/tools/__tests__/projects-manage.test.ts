import type { PersonalProject, TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import { projectsManage } from '../projects-manage.js'
import {
    TEST_IDS,
    createMockProject,
    extractStructuredContent,
    extractTextContent,
} from '../test-helpers.js'

// Mock the Todoist API
const mockTodoistApi = {
    addProject: jest.fn(),
    updateProject: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

describe('projects-manage tool', () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('creating a new project', () => {
        it('should create a project and return mapped result', async () => {
            const mockApiResponse = createMockProject({
                id: TEST_IDS.PROJECT_TEST,
                name: 'test-abc123def456-project',
                childOrder: 1,
                createdAt: '2024-01-01T00:00:00Z',
            })

            mockTodoistApi.addProject.mockResolvedValue(mockApiResponse)

            const result = await projectsManage.execute(
                { name: 'test-abc123def456-project' },
                mockTodoistApi,
            )

            // Verify API was called correctly
            expect(mockTodoistApi.addProject).toHaveBeenCalledWith({
                name: 'test-abc123def456-project',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Created project: test-abc123def456-project')
            expect(textContent).toContain(`id=${TEST_IDS.PROJECT_TEST}`)
            expect(textContent).toContain('Use tasks-add-multiple to add your first tasks')

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    project: expect.objectContaining({
                        id: TEST_IDS.PROJECT_TEST,
                        name: 'test-abc123def456-project',
                    }),
                    operation: 'created',
                }),
            )
        })

        it('should handle different project properties from API', async () => {
            const mockApiResponse = createMockProject({
                id: 'project-456',
                name: 'My Blue Project',
                color: 'blue',
                isFavorite: true,
                isShared: true,
                parentId: 'parent-123',
                viewStyle: 'board',
                childOrder: 2,
                description: 'A test project',
                createdAt: '2024-01-01T00:00:00Z',
            })

            mockTodoistApi.addProject.mockResolvedValue(mockApiResponse)

            const result = await projectsManage.execute({ name: 'My Blue Project' }, mockTodoistApi)

            expect(mockTodoistApi.addProject).toHaveBeenCalledWith({ name: 'My Blue Project' })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Created project: My Blue Project')
            expect(textContent).toContain('id=project-456')
            expect(textContent).toContain('Use sections-manage to organize this project')
        })
    })

    describe('updating an existing project', () => {
        it('should update a project when id is provided', async () => {
            const mockApiResponse: PersonalProject = {
                url: 'https://todoist.com/projects/existing-project-123',
                id: 'existing-project-123',
                parentId: null,
                isDeleted: false,
                updatedAt: '2025-08-13T22:10:30.000000Z',
                childOrder: 1,
                description: '',
                isCollapsed: false,
                canAssignTasks: false,
                color: 'red',
                isFavorite: false,
                isFrozen: false,
                name: 'Updated Project Name',
                viewStyle: 'list',
                isArchived: false,
                inboxProject: false,
                isShared: false,
                createdAt: '2024-01-01T00:00:00Z',
                defaultOrder: 0,
            }

            mockTodoistApi.updateProject.mockResolvedValue(mockApiResponse)

            const result = await projectsManage.execute(
                { id: 'existing-project-123', name: 'Updated Project Name' },
                mockTodoistApi,
            )

            expect(mockTodoistApi.updateProject).toHaveBeenCalledWith('existing-project-123', {
                name: 'Updated Project Name',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Updated project: Updated Project Name')
            expect(textContent).toContain('id=existing-project-123')
            expect(textContent).toContain('Use projects-list to see all projects')
        })
    })

    describe('error handling', () => {
        it('should propagate API errors', async () => {
            const apiError = new Error('API Error: Project name is required')
            mockTodoistApi.addProject.mockRejectedValue(apiError)

            await expect(projectsManage.execute({ name: '' }, mockTodoistApi)).rejects.toThrow(
                'API Error: Project name is required',
            )
        })
    })
})
