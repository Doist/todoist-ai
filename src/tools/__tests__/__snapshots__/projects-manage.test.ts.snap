// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`projects-manage tool creating a new project should create a project and return mapped result 1`] = `
"Created project: test-abc123def456-project • id=6cfCcrrCFg2xP94Q
Next:
- Use sections-manage to organize this project with sections
- Use tasks-add-multiple to add your first tasks
- Use overview with projectId to see updated project structure."
`;

exports[`projects-manage tool creating a new project should handle different project properties from API 1`] = `
"Created project: My Blue Project • id=project-456
Next:
- Use sections-manage to organize this project with sections
- Use tasks-add-multiple to add your first tasks
- Use overview with projectId to see updated project structure."
`;

exports[`projects-manage tool updating an existing project should update a project when id is provided 1`] = `
"Updated project: Updated Project Name • id=existing-project-123
Next:
- Use overview with projectId=existing-project-123 to see project structure
- Use projects-list to see all projects with updated name
- Use tasks-list-for-container to review existing tasks"
`;
